proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+```[UnscopedRef] public System.Span<T> AsSpan()``` method on the converted variable. 
+
+Result of a fixed-size buffer conversion to ```System.ReadOnlySpan<T>``` is a value equaivalent to invoking
+```[UnscopedRef] public readonly System.ReadOnlySpan<T> AsReadOnlySpan()``` method on the converted variable. 
+
+For example:
+``` C#
+void M1(Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // Ok, equivalent to `System.Span<int> b = x.AsSpan()`
+}
+
+void M2(in Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // An error, no conversion to System.Span<int>
+}
+
+Buffer10<int> GetBuffer() => default;
+
+void M3()
+{
+    System.ReadOnlySpan<int> a = GetBuffer(); // An error, no conversion to System.ReadOnlySpan<int>
+    System.Span<int> b = GetBuffer(); // An error, no conversion to System.Span<int>
+}
+```
+
+Alternatively, we could say that there is an implicit conversion from expression of a fixed-size buffer type to the following types
+(regardless of vaariable/value classification):
+- ```System.Span<T>```
+- ```System.ReadOnlySpan<T>```
+
+But then, converting a readonly variable to ```System.Span<T>``` or converting a value to either type is a ref-safety error.
+
+
+For example:
+``` C#
+void M1(Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // Ok, equivalent to `System.Span<int> b = x.AsSpan()`
+}
+
+void M2(in Buffer10<int> x)
+{
+    System.ReadOnlySpan<int> a = x; // Ok, equivalent to `System.ReadOnlySpan<int> a = x.AsReadOnlySpan()`
+    System.Span<int> b = x; // An error, readonly mismatch
+}
+
+Buffer10<int> GetBuffer() => default;
+
+void M3()
+{
+    System.ReadOnlySpan<int> a = GetBuffer(); // An error, ref-safety
+    System.Span<int> b = GetBuffer(); // An error, ref-safety
+}
+```
+
+With this approach, less context information is needed for conversion classification. This is going to significantly simplify implementation. 
+
+
+### List patterns
+
+[List patterns](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md) will not be supported for instances
+of fixed-size buffer types.
+
+### MemoryMarshal.CreateReadOnlySpan
+
+We might want to specially recognize ```System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan``` method and allow
+passing a reference to a readonly location as its first argument. This should allow users to define their own fixed-size buffer
+types. For example, in order to be able to share a type across multiple assemblies and also include custom APIs into its definition.
+This might also help framework to define our shared fixed-size buffer types.
```

Comment:
Developers can do this today with:
```C#
MemoryMarshal.CreateReadOnlySpan(ref Unsafe.AsRef(in location), length);
```
But this is also an example of an API that's motivation for allowing `ref readonly` on parameters; we would change CreateReadOnlySpan to accept a `ref readonly` instead of a `ref`, existing usage would continue to work, but it would also then work for creating a span around readonly fields.